src/Compiler/CompilerOptions.cpp

```diff
@@ -709,21 +888,160 @@ void delCompilerConfig(std::string k, std::vector<std::string> v) {
+
+  llvm::SmallString<8> instDir(kInstPath);
+  llvm::sys::path::append(instDir, "lib");
+  return llvm::StringRef(instDir).str();
+}
+
+// onnx-mlir currently requires llvm tools llc and opt and they are assumed
+// to be under llvm-project/build/bin. This doesn't work with the case where
+// llvm-project has been installed system wide (typically under /usr/local/...)
+// and its source has been removed.
+//
+// To account for this scenario, we first search for the tools in the same
+// directory where onnx-mlir is run. If they are found, it means both onnx-mlir
+// and llvm-project have been installed system wide under the same directory,
+// so we get them from that directory (typically /usr/local/bin). Otherwise,
+// at least one of onnx-mlir and llvm-project has not been installed system
+// wide. In this case, getToolPath returns the fallback directory where llvm
+// is built which is typically llvm-project/build/bin.
+//
+// Note that this will not work if both onnx-mlir and llvm-project have been
+// installed system wide but to different places and their sources have been
+// removed. So we force CMAKE_INSTALL_PREFIX to be the same as that of
+// llvm-project.
+//
+// If the flag is true, getToolPath will simply return the path detected by
+// cmake at compile time. This is used for system wide tools such as cc, ld, ar,
+// etc. Note that this means the path is valid only on the system where
+// onnx-mlir is built. If onnx-mlir is subsequently run on a system that does
+// not have these tools installed in the "standard" places, it will fail.
+//
+// Setting flag = true is also used to simply look up non-path config such
+// as lrodataScript.
+std::string getToolPath(
+    const std::string &tool, bool flag /*false by default*/) {
+
+  if (!flag) {
+    std::string execDir = llvm::sys::path::parent_path(getExecPath()).str();
+    llvm::SmallString<8> toolPath(execDir);
+    llvm::sys::path::append(toolPath, tool);
+    std::string p = llvm::StringRef(toolPath).str();
+    if (llvm::sys::fs::can_execute(p))
+      return p;
+  }
+
+  return toolPathMap.at(tool);
+}
+
+// This function must be called after llvm::cl::ParseCommandLineOptions
+// to create whatever options related compiler configuration states
+// based on the parsed options. It can also check for option consistency.
+//
+// The reason we don't put llvm::cl::ParseCommandLineOptions and
+// initCompilerConfig in a single function is that according to llvm doc
+// llvm::cl::ParseCommandLineOptions should be called from main.
+//
+// The caller can pass in a "hook" function to perform special handling
+// of the options after they are parsed. For example, onnx-mlir must
+// have the input filename specified and it derives the output filename
+// from the input filename if -o is not specified. onnx-mlir-opt, on the
+// other hand, defaults to stdin/stdout if the input/output filenames
+// are not specified.
+void initCompilerConfig(void (*hook)(void) /* nullptr by default */) {
```

Comment:
suggestion: skip the hook argument and instead change the caller to call the hook after calling initCompilerConfig - that seems a little bit simpler and I'm skeptical whether we'll use the hook mechanism for many other things in the future if you can only pass a single hook

Reply:
Given the fact that the hook is called at the end, I would lean toward @sorenlassen suggestion too.